invesalius/data/elfin_processing.py

```diff
@@ -0,0 +1,182 @@
+#    detalhes.
+#--------------------------------------------------------------------------
+import numpy as np
+import cv2
+from time import time
+
+import invesalius.data.transformations as tr
+import invesalius.data.coregistration as dcr
+import invesalius.constants as const
+
+
+class KalmanTracker:
+    def __init__(self,
+                 state_num=2,
+                 cov_process=0.001,
+                 cov_measure=0.1):
+
+        self.state_num = state_num
+        measure_num = 1
+
+        # The filter itself.
+        self.filter = cv2.KalmanFilter(state_num, measure_num, 0)
+
+        self.state = np.zeros((state_num, 1), dtype=np.float32)
+        self.measurement = np.array((measure_num, 1), np.float32)
+        self.prediction = np.zeros((state_num, 1), np.float32)
+
+
+        self.filter.transitionMatrix = np.array([[1, 1],
+                                                 [0, 1]], np.float32)
+        self.filter.measurementMatrix = np.array([[1, 1]], np.float32)
+        self.filter.processNoiseCov = np.array([[1, 0],
+                                                [0, 1]], np.float32) * cov_process
+        self.filter.measurementNoiseCov = np.array( [[1]], np.float32) * cov_measure
+
+    def update_kalman(self, measurement):
+        self.prediction = self.filter.predict()
+        self.measurement = np.array([[np.float32(measurement[0])]])
+
+        self.filter.correct(self.measurement)
+        self.state = self.filter.statePost
+
+
+class TrackerProcessing:
+    def __init__(self):
+        self.coord_vel = []
+        self.timestamp = []
+        self.velocity_vector = []
+        self.kalman_coord_vector = []
+        self.velocity_std = 0
+
+        self.tracker_stabilizers = [KalmanTracker(
+            state_num=2,
+            cov_process=0.001,
+            cov_measure=0.1) for _ in range(6)]
+
+    def kalman_filter(self, coord_tracker):
+        kalman_array = []
+        pose_np = np.array((coord_tracker[:3], coord_tracker[3:])).flatten()
+        for value, ps_stb in zip(pose_np, self.tracker_stabilizers):
+            ps_stb.update_kalman([value])
+            kalman_array.append(ps_stb.state[0])
+        coord_kalman = np.hstack(kalman_array)
+
+        self.kalman_coord_vector.append(coord_kalman[:3])
+        if len(self.kalman_coord_vector) < 20: #avoid initial fluctuations
+            coord_kalman = coord_tracker
+            print('initializing filter')
+        else:
+            del self.kalman_coord_vector[0]
+
+        return coord_kalman
+
+    def estimate_head_velocity(self, coord_vel, timestamp):
+        coord_vel = np.vstack(np.array(coord_vel))
+        coord_init = coord_vel[:int(len(coord_vel)/2)].mean(axis=0)
+        coord_final = coord_vel[int(len(coord_vel)/2):].mean(axis=0)
+        velocity = (coord_final - coord_init)/(timestamp[-1]-timestamp[0])
+        distance = (coord_final - coord_init)
+
+        return velocity, distance
+
+    def versors(self, init_point, final_point):
```

Comment:
What is this 70? I guess it is an arbitrary scale number, but would be good to explicitely document it.